core/llm_google.go

```diff
@@ -164,91 +122,193 @@ func (c *GenaiClient) SendQuery(ctx context.Context, history []ModelMessage, too
 			genaiHistory = append(genaiHistory, content)
 		}
 	}
-	if len(model.SystemInstruction.Parts) == 0 {
-		model.SystemInstruction = nil
-	}
 
-	// Pop last message from history for SendMessage
-	if len(genaiHistory) == 0 {
-		return nil, fmt.Errorf("no user prompt")
+	if len(systemInstruction.Parts) == 0 {
+		systemInstruction = nil
 	}
 
-	userMessage := genaiHistory[len(genaiHistory)-1]
-	genaiHistory = genaiHistory[:len(genaiHistory)-1]
+	return genaiHistory, systemInstruction, nil
+}
 
-	chat := model.StartChat()
-	chat.History = genaiHistory
+// Define callbacks passed to processStreamResponse for handling stream events
+// We isolate the core stream iteration logic from the application logic (logging, metrics, final accumulation)
+type textPartHandler func(ctx context.Context, text string) error
+type tokenUsageHandler func(ctx context.Context, usageMeta *genai.UsageMetadata, usageSummary *LLMTokenUsage) error
 
-	stream := chat.SendMessageStream(ctx, userMessage.Parts...)
+func (c *GenaiClient) processStreamResponse(
+	ctx context.Context,
+	stream *genai.GenerateContentResponseIterator,
+	onTextPart textPartHandler,
+	onTokenUsage tokenUsageHandler,
+) (content string, toolCalls []ToolCall, tokenUsage LLMTokenUsage, err error) {
+
+	content = ""
+	toolCalls = []ToolCall{}
+	tokenUsage = LLMTokenUsage{}
 
-	var content string
-	var toolCalls []ToolCall
-	var tokenUsage LLMTokenUsage
 	for {
-		res, err := stream.Next()
-		if err != nil {
-			if errors.Is(err, iterator.Done) {
+		res, streamErr := stream.Next()
```

Comment:
is it because of the `err` named return that you changed `err` to `streamErr` ? Should be fine to keep it `err`, especially since you are always specifying the return params at each `return` statement.